internal/corazawaf/rule.go

```diff
@@ -175,6 +198,32 @@ func (r *Rule) doEvaluate(phase types.RulePhase, tx *Transaction, cache map[tran
 		rid = r.ParentID_
 	}
 
+	// TODO(anuraaga): This is effectively lazily computing the min phase of a rule with chain the first
+	// time we evaluate the rule. Instead, we should do this at parse time, but this will require a
+	// large-ish refactoring of the parser, which adds parent rules to a rule group before preparing
+	// the child rules. In the meantime, only evaluating this once should allow performance to be fine.
+	//
+	// chainMinPhase is the minimum phase among all the rules in which the chained rule may match.
+	// We evaluate the min possible phase for each rule in the chain and we take the minimum in common
+	// If we reached this point, it means that the parent rule already reached its min phase.
+	if multiphaseEvaluation && r.ParentID_ == 0 && r.HasChain && r.chainMinPhase == types.PhaseUnknown {
+		for c := r.Chain; c != nil; c = c.Chain {
+			singleChainedRuleMinPhase := types.PhaseUnknown
```

Comment:
While doing some manual tests with Envoy I found that we were trying to evaluate Chained rules too much, in certain circumstances bypassing also paranoia levels.
I refactored a bit the way we compute `chainMinPhase`. Now the chainMinPhase is the latest minPhase across all the chained rules.

For example:
```
SecRule REQUEST_URI "/chain_phase2" "id:10, phase:2, t:none, log, setvar:'tx.set1=1', chain"
	SecRule REQUEST_URI "/chain_phase2" "setvar:'tx.set2=2', chain"
		SecRule REQUEST_BODY "chain_phase2" "setvar:'tx.set3=3'" 
```
Here I think that REQUEST_BODY is the "bottle neck" of this chain, therefore it is the chainMinPhase. Without being able to evaluate it, the chain will never match.